// testcafe/lib/utils/get-options/quarantine.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Quarantine = exports.getQuarantineOptions = void 0;
const base_1 = __importDefault(require("./base"));
const quarantine_option_names_1 = __importDefault(require("../../configuration/quarantine-option-names"));
const types_1 = require("../../errors/types");
const runtime_1 = require("../../errors/runtime");
const DEFAULT_QUARANTINE_THRESHOLD = 3;
const DEFAULT_TEST_RUN_THRESHOLD = 5;
function _isQuarantineOption(option) {
    return Object.values(quarantine_option_names_1.default).includes(option);
}
function _validateQuarantineOptions(options) {
    const retryCount = options.retryCount || DEFAULT_TEST_RUN_THRESHOLD;
    const passCount = options.passCount || DEFAULT_QUARANTINE_THRESHOLD;
    if (passCount > retryCount)
        throw new runtime_1.GeneralError(types_1.RUNTIME_ERRORS.invalidRetryCountValue, passCount);
}
async function getQuarantineOptions(optionName, options) {
    if (typeof options === 'boolean')
        return true;
    const parsedOptions = await base_1.default(options, {
        skipOptionValueTypeConversion: true,
        async onOptionParsed(key, value) {
            if (!key || !value)
                throw new runtime_1.GeneralError(types_1.RUNTIME_ERRORS.optionValueIsNotValidKeyValue, optionName);
            return Number(value);
        }
    });
    if (Object.keys(parsedOptions).some(key => !_isQuarantineOption(key)))
        throw new runtime_1.GeneralError(types_1.RUNTIME_ERRORS.invalidQuarantineOption, optionName);
    _validateQuarantineOptions(parsedOptions);
    return parsedOptions;
}
exports.getQuarantineOptions = getQuarantineOptions;
class Quarantine {
    constructor() {
        this.attempts = [];
        this.testRunThreshold = DEFAULT_TEST_RUN_THRESHOLD;
        this.failedQuarantineThreshold = DEFAULT_QUARANTINE_THRESHOLD;
        this.passedQuarantineThreshold = DEFAULT_QUARANTINE_THRESHOLD;
    }
    getFailedAttempts() {
        return this.attempts.filter(errors => !!errors.length);
    }
    getPassedAttempts() {
        return this.attempts.filter(errors => errors.length === 0);
    }
    setPassedQuarantineThreshold(threshold) {
        this.passedQuarantineThreshold = threshold;
        this._setFailedThreshold();
    }
    setTestRunThreshold(threshold) {
        this.testRunThreshold = threshold;
        this._setFailedThreshold();
    }
    getNextAttemptNumber() {
        return this.attempts.length + 1;
    }
    isThresholdReached(extraErrors) {
        const { failedTimes, passedTimes } = this._getAttemptsResult(extraErrors);
        const failedThresholdReached = failedTimes >= this.failedQuarantineThreshold;
        const passedThresholdReached = passedTimes >= this.passedQuarantineThreshold;
        return failedThresholdReached || passedThresholdReached;
    }
    isFirstAttemptSuccessful(extraErrors) {
        const { failedTimes, passedTimes } = this._getAttemptsResult(extraErrors);
        return failedTimes === 0 && passedTimes > 0;
    }
    _getAttemptsResult(extraErrors) {
        let failedTimes = this.getFailedAttempts().length;
        let passedTimes = this.getPassedAttempts().length;
        if (extraErrors) {
            if (extraErrors.length)
                failedTimes += extraErrors.length;
            else
                passedTimes += 1;
        }
        return { failedTimes, passedTimes };
    }
    _setFailedThreshold() {
        if (this.testRunThreshold !== DEFAULT_TEST_RUN_THRESHOLD)
            this.failedQuarantineThreshold = this.testRunThreshold - this.passedQuarantineThreshold + 1;
        else
            this.failedQuarantineThreshold = DEFAULT_QUARANTINE_THRESHOLD;
    }
}
exports.Quarantine = Quarantine;
